README.md
# If On A Winters Night a Traveler

## Casual NLG GUI app

> Every time I sit down here I read, "it was a dark and stormy night . . ." and the impersonality of the incipit seems to open the passage from one world to the other, from the time and space of here and now to the time and space of the written word; I feel the thrill of a beginning that can be followed by multiple developments, inexhaustibly...
> -Italo Calvino, *If on a Winter's Night a Traveler*

![Screenshot](UI.png)

Easy to use GUI wrapper for running NLG models from HuggingFace locally.

Third party dependencies are PySimpleGUI and aitextgen

Provide a string of text as a prompt (or don't), modify the generator's parameters, and press "Generate."
Can generate up to 20 strings at a time of 750 word length without modification.
The more strings and the longer they are, the more time is required.

The models run on your CPU or CUDA compatible GPUs.
They can be quite large, so I deliberately restricted the list of available models, but you could easily expand it.
This app lets you choose where models are stored as well.

Please remember to use a virtual environment.

### Installation and Running
```
pip install -r requirements.txt
```
and
```
py main.py
```
#### What do I do if my CUDA capable GPU isn't working?
I think pip gets a CPU only version of PyTorch for aitextgen by default. 
Go to [this link](https://pytorch.org/get-started/locally) and follow the instructions on a clean virtual environment after already installing aitextgen, making sure to select CUDA 10.2/11.3 instead of CPU.


CHANGELOG.md
#--- 0.0.2 --- Feb 22, 2022 ---
Added GPU support

####--backend--
Some Thread related work for halting feature

#--- 0.0.1 --- Feb 17, 2022 ---

Project created

main.py
import PySimpleGUI as sg

import EventLoop
import TextGenWindow

sg.user_settings_filename(path=".")
window = TextGenWindow.TextGenWindow()

while True:
    if EventLoop.MainEventLoop.main_event_loop(window) == -1:
        if window.result_window_active:
            window.result_window.close()
        if window.settings_window_active:
            window.settings_window.close()
        break

    if window.result_window_active:
        if EventLoop.ResultEventLoop.r_event_loop(window) == -1:
            window.result_window_active = False
            window.result_window = None

    if window.settings_window_active:
        if EventLoop.SettingsEventLoop.settings_event_loop(window) == -1:
            window.settings_window_active = False
            window.settings_window = None

window.close()


EventLoop.py
import datetime as dt
from tkinter import Tk

import PySimpleGUI as sg

import Layout
import Models
from TextGenWindow import TextGenWindow
from util import save_text_to_file


class MainEventLoop:
    @staticmethod
    def main_event_loop(window: TextGenWindow):
        event, values = window.read(timeout=100)

        generator = window.generator

        if event == sg.WIN_CLOSED or event == 'Exit' or event == '-EXIT-':
            if window.result_window:
                window.result_window.close()
            return -1

        MainEventLoop.text_generation_behavior(event, values, window, generator)
        MainEventLoop.model_swapping_behavior(event, values, window, generator)

        if event == '-IN-' and values['-IN-'] and values['-IN-'][-1] not in ('0123456789.-'):
            window['-IN-'].update(values['-IN-'][:-1])

        if event == "Settings":
            SettingsEventLoop.settings_popup(window)

        if event == "-HALT-":
            # print(generator.threads)
            # [print(dir(thread)) for thread in generator.threads]
            # [print(thread.is_alive()) for thread in generator.threads]
            # [print(thread.daemon) for thread in generator.threads]
            # TODO
            pass

    @staticmethod
    def text_generation_behavior(event, values, window, generator):
        if event == "-GENERATE-":
            window.set_cursor("watch")
            window['-GENERATE-'].update(disabled=True)
            window['-MODEL_SELECT-'].update(disabled=True)
            if not window['-HALT-'].visible:
                window['-HALT-'].update(visible=True)
            window['-HALT-'].update(disabled=False)

            temperature = float(values['-TEMPERATURE-'])
            prompt = values["-PROMPT-"]
            iter_count = values['-SPIN-']
            word_count = int(values["-WORD_LENGTH-"]) + len(prompt)

            generator.generate(prompt=prompt, length=word_count, n=iter_count, temperature=temperature)

            window['-RESULTS_ALERT-'].update(value=f"Results pending...")
            if not window['-RESULTS_ALERT-'].visible:
                window['-RESULTS_ALERT-'].update(visible=True)

        if event == "-TEXT_GENERATED-":
            window.set_cursor("")
            txt = f"Model: {Models.current_model}\nPrompt: {generator.prompt}\nTemperature: {generator.temperature}\n\
Length: {generator.word_count}\n"

            txt += "-" * 50 + '\n'
            for i in range(len(values['-TEXT_GENERATED-'])):
                txt += f'{i + 1} of {len(values["-TEXT_GENERATED-"])}' + '\n'
                txt += values['-TEXT_GENERATED-'][i] + '\n'
                txt += "-" * 50 + '\n'

            fname = f"results/{dt.datetime.now().strftime('%d.%m.%y, %H.%M')}.txt"
            save_text_to_file(fname, txt)
            window.ding()
            window['-RESULTS_ALERT-'].update(value=f"Results saved to:\n {fname}")
            if not window['-RESULTS_ALERT-'].visible:
                window['-RESULTS_ALERT-'].update(visible=True)
            window['-GENERATE-'].update(disabled=False)
            window['-HALT-'].update(disabled=True)
            window['-MODEL_SELECT-'].update(disabled=False)

            # LAUNCHES A WINDOW
            if values['-RESULT_WINDOW_TOGGLE-']:
                ResultEventLoop.results_popup(window, generator, values['-TEXT_GENERATED-'], fname)

    @staticmethod
    def model_swapping_behavior(event, values, window, generator):
        if event == "-MODEL_SELECT-":
            if values['-MODEL_SELECT-'][0] == Models.current_model:
                window['-MODEL_SWAP-'].update(disabled=True)
            elif generator.model_loaded and window['-MODEL_SWAP-'].Disabled:
                window['-MODEL_SWAP-'].update(disabled=False)

        if event == "-MODEL_LOADED-":
            window.ding()
            window.set_cursor("")

            window['-MODEL_SELECT-'].update(disabled=False)
            window['-MODEL_SWAP-'].update(disabled=False)
            window['-GENERATE-'].update(disabled=False)
            window.set_cursor('')
            window["-MODEL_LABEL-"].update(Models.current_model)

            # set the '-MODEL_SELECT-' to not equal the current model
            new_possible_model_index = Models.models.index(Models.current_model) + 1
            if new_possible_model_index > len(Models.models) - 1:
                new_possible_model_index = 0
            window['-MODEL_SELECT-'].update(set_to_index=new_possible_model_index)

            if sg.user_settings_get_entry("on_gpu"):
                window['-MODEL_LOCATION-'].update("Model on GPU")
            else:
                window['-MODEL_LOCATION-'].update("Model on CPU")

        if event == "-MODEL_SWAP-":
            window.set_cursor("watch")
            window['-MODEL_SWAP-'].update(disabled=True)
            window['-MODEL_SELECT-'].update(disabled=True)
            window['-GENERATE-'].update(disabled=True)
            if not window['-HALT-'].visible:
                window['-HALT-'].update(visible=True)
            window['-HALT-'].update(disabled=False)

            window.set_cursor('wait')
            Models.current_model = values["-MODEL_SELECT-"][0]
            generator.load_model()
            window["-MODEL_LABEL-"].update("Model swap pending...")
        pass


class ResultEventLoop:
    @staticmethod
    def r_event_loop(main_window):
        r_window = main_window.result_window
        event, values = r_window.read(timeout=100)

        if event == sg.WIN_CLOSED or event == 'Exit' or event == '-EXIT-':
            r_window.close()
            return -1

        if event[:len('-COPY_RESULT')] == '-COPY_RESULT':
            n = int(event[len('-COPY_RESULT') + 1:-1])
            copy2clip(r_window[f'-RESULT_{n}_TEXT-'].DisplayText)
            return

        if event[:len('-SUBMIT_RESULT')] == '-SUBMIT_RESULT':
            n = int(event[len('-SUBMIT_RESULT') + 1:-1])
            new_prompt = values[f'-RESULT_{n}_TEXT-']
            main_window.new_result_set_from_result(new_prompt)
            return

    @staticmethod
    def results_popup(window, gen, results, fname):
        window.result_window_active = True
        l = Layout.ResultsLayout.generate_results_layout(gen, results, fname)
        if window.result_window:
            window.result_window.close()
        window.result_window = sg.Window("Results", layout=l, resizable=True, finalize=True)


class SettingsEventLoop:
    @staticmethod
    def settings_event_loop(main_window):
        s_window = main_window.settings_window
        event, values = s_window.read(timeout=100)

        if event == sg.WIN_CLOSED or event == 'Exit' or event == '-EXIT-':
            s_window.close()
            return -1

        if event == '-SETTINGS_SAVE-':
            yn = sg.PopupYesNo("This will close the application, is that ok?")
            if yn == 'Yes':
                sg.user_settings_set_entry("cache_dir", values['-SETTINGS_CACHE_DIR_INPUT-'])
                sg.user_settings_set_entry("on_gpu", values["-SETTINGS_LOAD_TO_GPU-"])
                s_window.close()
                main_window.close()
                return -1
            if yn == "No":
                return

    @staticmethod
    def settings_popup(window):
        window.settings_window_active = True
        l = Layout.SettingsLayout.generate_settings_layout()
        if window.settings_window:
            window.settings_window.close()
        window.settings_window = sg.Window("Settings", layout=l, finalize=True, size=(700,200))


def copy2clip(txt):
    r = Tk()
    r.withdraw()
    r.clipboard_clear()
    r.clipboard_append(txt)
    r.update()  # now it stays on the clipboard after the window is closed
    r.destroy()


Generator.py
import PySimpleGUI as sg

import Models
from aitextgen import aitextgen
from torch.cuda import is_available


class Generator:
    model = Models.current_model
    model_loaded = False
    threads = []

    def generate(self, prompt: str, length: int, n: int, temperature: float, ):
        self.prompt = prompt
        self.word_count = length
        self.n = n
        self.temperature = temperature

        self.threads.append(self.window.perform_long_operation(lambda:
                                                               self._ai.generate(prompt=prompt,
                                                                                 max_length=length,
                                                                                 temperature=temperature,
                                                                                 return_as_list=True,
                                                                                 n=n),
                                                               "-TEXT_GENERATED-"
                                                               ))

    def get_model(self):
        gpu = False
        if sg.user_settings_get_entry("on_gpu"):
            if is_available():
                gpu = True
            else:
                sg.PopupAnnoying("Tried to load model to GPU but CUDA wasn't available.")

        if sg.user_settings_get_entry("cache_dir") is None:
            raise Exception()
        else:
            self._ai = aitextgen(
                model=Models.current_model,
                to_gpu=gpu,
                cache_dir=sg.user_settings_get_entry("cache_dir")
            )
            self.model_loaded = True

    def load_model(self):
        self.model_loaded = False
        self.window.perform_long_operation(lambda:
                                           self.get_model(),
                                           "-MODEL_LOADED-"
                                           )

    def __init__(self, window):
        self.window = window
        self._ai = None
        self.prompt = None
        self.word_count = None
        self.n = None
        self.temperature = None
        self.load_model()


TextGenWindow.py
import os

import PySimpleGUI as sg

import Generator
import Layout


class TextGenWindow(sg.Window):
    def __init__(self):
        super().__init__("Text Generator", Layout.MainLayout.layout, finalize=True)
        self.result_window = None
        self.settings_window = None
        self.result_window_active = False
        self.settings_window_active = False

        if sg.user_settings_get_entry("cache_dir") is None:
            new_dir = sg.PopupGetFolder(
                title="Cache size warning",
                message="Would you like to change the folder the models are stored in?\n"
                        "By default they will be saved in a folder alongside the executable called 'aitextgen'.\n"
                        "Close or cancel for default behavior.\n"
                        "This window only appears if the 'aitextgen' folder is missing and there is no set alternative."
            )
            if new_dir is not None:
                sg.user_settings_set_entry("cache_dir", new_dir)
            else:
                default_dir = os.path.abspath("aitextgen/")
                sg.user_settings_set_entry("cache_dir", default_dir)

        if sg.user_settings_get_entry("on_gpu") is None:
            ans = sg.PopupYesNo("Run models on GPU?")
            if ans == "Yes":
                ans = True
            else:
                ans = False

            sg.user_settings_set_entry("on_gpu", ans)

        self.generator = Generator.Generator(self)

    def new_result_set_from_result(self, new_prompt):
        self['-PROMPT-'].update(value=new_prompt)

        self['-GENERATE-'].click()
        pass


Layout.py
from textwrap import wrap

import PySimpleGUI as sg

import Models


def place(elem):
    """
    Places element provided into a Column element so that its placement in the layout is retained.
    :param elem: the element to put into the layout
    :return: A column element containing the provided element
    """
    return sg.Column([[elem]], pad=(0, 0))


sg.theme('DarkAmber')


class MainLayout:
    layout = [
        [
            sg.Menu([
                ["File", ["Settings", "Exit"]]
            ])
        ],
        [
            sg.Text('Currently loaded model:'),
            sg.Text("Model pending...", key="-MODEL_LABEL-"),
            sg.Text("", key="-MODEL_LOCATION-"),
        ],
        [
            sg.Listbox(
                Models.models, enable_events=True, key="-MODEL_SELECT-", disabled=True, size=(25, 5),
                select_mode="LISTBOX_SELECT_MODE_SINGLE"
            ),
            sg.Btn("Swap models", key="-MODEL_SWAP-", disabled=True)
        ],
        [
            sg.Text('Prompt string:')
        ],
        [
            sg.InputText("If on a winter's night a traveler", key="-PROMPT-"), sg.Text("Strings to generate"),
            sg.Spin([x for x in range(1, 21)], key='-SPIN-', size=(4, 1))
        ],
        [
            sg.Text("Temperature:"),
            sg.Input(key='-TEMPERATURE-', enable_events=True, default_text='1.0', size=(6, 1)),
        ],
        [
            sg.Text("Word length:"),
            sg.Slider((50, 750), key='-WORD_LENGTH-', enable_events=True, orientation="horizontal", expand_x=True)
        ],
        [
            sg.Button('Generate', key="-GENERATE-", disabled=True),
            sg.Check("Open result window", key="-RESULT_WINDOW_TOGGLE-", default=True),
            place(sg.Text("", visible=False, key="-RESULTS_ALERT-")),
            place(sg.Button('Halt', visible=False, key="-HALT-", tooltip="TODO")),
        ],
    ]


class ResultsLayout:
    @staticmethod
    def generate_results_layout(gen, results, fname):
        return [
            [ResultsLayout._header(gen.temperature, gen.word_count)],
            [ResultsLayout._generate_result_table(results)],
            [ResultsLayout._footer(fname)]
        ]

    @staticmethod
    def _header(temp, length):
        return [
                   sg.Text("Batch Info"),
                   sg.Text("Temperature:"),
                   sg.Text(temp, key="-RESULTS_TEMPERATURE-"),
                   sg.Text("Length:"),
                   sg.Text(length, key="-RESULTS_LENGTH-"),
               ],

    @staticmethod
    def _generate_result_table(results):
        result_table = []
        n = len(results)

        for i in range(n):
            result_table.extend(ResultsLayout._prepare_result_row(i, n, results[i]))

        return sg.Frame("Results",
                        [[sg.Column(
                            layout=result_table,
                            element_justification='left',
                            scrollable=True,
                            vertical_scroll_only=True,
                            expand_y=True,
                            expand_x=True,
                            key="-RESULTS_COLUMN-"
                        )]],
                        key="-RESULTS_FRAME-",
                        size=(511, 600),
                        expand_x=True,
                        expand_y=True,
                        )

    @staticmethod
    def calculate_row_height(text):
        y_size = 2
        if "\n" in text:
            y_size += text.count("\n")
        return y_size

    @staticmethod
    def _prepare_result_row(n, m, result):
        result = "\n".join(wrap(result, 50))
        y_size = ResultsLayout.calculate_row_height(result)

        rr = [
            [
                sg.Text(f"{n + 1} of {m}", key=f'-RESULT_{n}_POSITION-'),
            ],
            [
                sg.Multiline(
                    f"{result.strip()}",
                    key=f'-RESULT_{n}_TEXT-',
                    size=(45, y_size),
                    disabled=True,
                    background_color=sg.theme_background_color(),
                    no_scrollbar=True,
                    border_width=0,
                ),
                sg.Column(
                    layout=[
                        [sg.Button("Copy", key=f'-COPY_RESULT_{n}-')],
                        [sg.Button("Submit\nas prompt", key=f'-SUBMIT_RESULT_{n}-', size=(7, 2))]
                    ],
                    element_justification="center"
                )
            ],
            [
                sg.HorizontalSeparator()
            ],
        ]
        return rr

    @staticmethod
    def _footer(fname):
        f = [
            sg.Text("Results written to"),
            sg.Text(fname, key="-RESULTS_FILE_LOCATION-")
        ]

        return f


class SettingsLayout:
    @staticmethod
    def generate_settings_layout():
        return [
            [
                sg.Text("Model cache directory:"),
            ],
            [
                sg.Input(
                    key='-SETTINGS_CACHE_DIR_INPUT-',
                    readonly=True,
                    default_text=sg.user_settings_get_entry("cache_dir"),
                    expand_x=True,
                    text_color='black'
                ),
                sg.FolderBrowse("Select", target="-SETTINGS_CACHE_DIR_INPUT-")
            ],
            [
                sg.HorizontalSeparator()
            ],
            [
                sg.Text("Load model to GPU:"),
                sg.Checkbox("Yes", key='-SETTINGS_LOAD_TO_GPU-', default=sg.user_settings_get_entry("on_gpu"))
            ],
            [
                sg.HorizontalSeparator()
            ],
            [
                sg.Column(
                    [
                        [
                            sg.Button("Save settings", key='-SETTINGS_SAVE-'),
                        ],
                    ],
                    justification='right',
                )
            ],
        ]
